### PR TITLE
fix: components not being removed on `ComponentRemoved` message

### PR DESCRIPTION
**Pull request by pbosio**
Pretty much every `IEntityComponent` was not being removed when receiving `ComponentRemoved` message from kernel (except from shapes and pointer events)

*added code to try to remove component by name when message is received
*modified `BaseComponent` class to ensure that every inheritor must set `componentName` property
*added class name to components that didn't have it

the following ecs script as an example, show billboard component is never removed
```
let active = true
const e1 = new Entity();
e1.addComponent(new Transform({ position: new Vector3(16 * 0.5, 1, 9), scale: new Vector3(16, 1, 14) }))
e1.addComponent(new BoxShape()).withCollisions = false
e1.addComponent(new Billboard())

Input.instance.subscribe("BUTTON_DOWN", ActionButton.PRIMARY, false, ev => {
    if (active) {
        e1.removeComponent(Billboard)
        log(`remove billboard`)
    }
    else {
        e1.addComponent(new Billboard())
        log(`create billboard`)
    }
    active = !active
})

engine.addEntity(e1)
```

### CODE REVIEW

**Review by BrianAmadori — CHANGES_REQUESTED** [review]
The implementation looks good, however we shouldn't be covering this sensible feature with tests?

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Animator/DCLAnimator.cs [hunk 1]
```diff
@@ -43,6 +43,8 @@ public class DCLAnimationState
         Dictionary<string, AnimationClip> clipNameToClip = new Dictionary<string, AnimationClip>();
         Dictionary<AnimationClip, AnimationState> clipToState = new Dictionary<AnimationClip, AnimationState>();
 
+        public override string componentName => "animator";
```
Comment (Suduck) [inline unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Animator/DCLAnimator.cs:46]: We should always uses camelCase or Snake, so this  should be Animator or the rest should be audioSource

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Components/AvatarAttach/AvatarAttachComponent.cs [hunk 2]
```diff
@@ -56,7 +56,7 @@ IEnumerator IComponent.ApplyChanges(BaseModel newModel)
 
         void IComponent.RaiseOnAppliedChanges() { }
 
-        Transform IMonoBehaviour.GetTransform() => handler.entity?.gameObject.transform;
+        Transform IMonoBehaviour.GetTransform() => null;
```
Comment (Suduck) [inline unity-renderer/Assets/Scripts/MainScripts/DCL/Components/AvatarAttach/AvatarAttachComponent.cs:59]: Is this change expected?

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Audio/DCLAudioSource.cs [hunk 3]
```diff
@@ -31,6 +31,8 @@ public class Model : BaseModel
         public long playedAtTimestamp = 0;
         private bool isOutOfBoundaries = false;
 
+        public override string componentName => "AudioSource";
```
Comment (Suduck) [inline unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Audio/DCLAudioSource.cs:34]: This too

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Audio/DCLAudioStream.cs [hunk 4]
```diff
@@ -28,6 +28,8 @@ public class Model : BaseModel
         private bool isDestroyed = false;
         private Model prevModel = new Model();
 
+        public override string componentName => "AudioStream";
```
Comment (Suduck) [inline unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Audio/DCLAudioStream.cs:31]: This too

**Review by Suduck — APPROVED** [review]
Good job! Small change requsted. Also, Do you consider that we should move all the names to constants ? Maybe in the CLASS_ID file? I let that decision to you!

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Components/AvatarAttach/AvatarAttachComponent.cs (hunk 2 above)
Comment (pbosio) [reply unity-renderer/Assets/Scripts/MainScripts/DCL/Components/AvatarAttach/AvatarAttachComponent.cs:59]: yes, `IMonoBehaviour.GetTransform` is only used for "pointer events" components so is not necessary to set something there

**Comment by pbosio** [comment]
> Good job! Small change requsted. Also, Do you consider that we should move all the names to constants ? Maybe in the CLASS_ID file? I let that decision to you!

I don't think we will gain something by moving those name to constants in other file since there is no hardcoded string in our project making any reference to them and I don't think it would mitigate the possibility of any unintended error. So I would prefer to keep them there explicitly in every component class

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Components/AvatarAttach/AvatarAttachComponent.cs (hunk 2 above)
Comment (pbosio) [reply unity-renderer/Assets/Scripts/MainScripts/DCL/Components/AvatarAttach/AvatarAttachComponent.cs:59]: Also, this component does not create any `GameObject` so it makes no sense setting something there :P

**Comment by pravusjif** [comment]
> > Good job! Small change requsted. Also, Do you consider that we should move all the names to constants ? Maybe in the CLASS_ID file? I let that decision to you!
> 
> I don't think we will gain something by moving those name to constants in other file since there is no hardcoded string in our project making any reference to them and I don't think it would mitigate the possibility of any unintended error. So I would prefer to keep them there explicitly in every component class

IT's true that many other component classes not modified in this PR already have the componentName override like `public override string componentName => "NFT Shape";`, In the future we can change that on all components to avoid memory fragmentation by instantiating a new string every time the `componentName` is used, right now it may not be needed

File: unity-renderer/Assets/Scripts/MainScripts/DCL/WorldRuntime/ParcelScene.cs [hunk 5]
```diff
@@ -684,6 +684,32 @@ private void RemoveEntityComponent(IDCLEntity entity, string componentName)
                         }
                     }
                     return;
+                default:
```
Comment (pravusjif) [inline unity-renderer/Assets/Scripts/MainScripts/DCL/WorldRuntime/ParcelScene.cs:687]: With this new `default` logic, do the `OnClick.NAME`, `OnPointerDown.NAME`, `OnPointerUp.NAME`, `OnPointerHoverEnter.NAME`, `OnPointerHoverExit.NAME` cases of that `switch` still make sense? couldn't those work fine with the new `default` as well?

Comment (pbosio) [reply unity-renderer/Assets/Scripts/MainScripts/DCL/WorldRuntime/ParcelScene.cs:687]: yes and no :P
having those as switch cases will make it faster than the default case which iterate through the component list to find the component to remove.
Ideally, for better performance, we should add a case in that switch for every component and the default case is mostly to handle the components that we forgot to add there

**Comment by pbosio** [comment]
> > > Good job! Small change requsted. Also, Do you consider that we should move all the names to constants ? Maybe in the CLASS_ID file? I let that decision to you!
> > 
> > 
> > I don't think we will gain something by moving those name to constants in other file since there is no hardcoded string in our project making any reference to them and I don't think it would mitigate the possibility of any unintended error. So I would prefer to keep them there explicitly in every component class
> 
> It's true that many other component classes not modified in this PR already have the componentName override like `public override string componentName => "NFT Shape";`, In the future we can move those to CONST on all components to avoid memory fragmentation by instantiating a new string every time the `componentName` is used, right now it may not be needed

hopefully, in the future, we won't need that component name at all :)

**Review by pravusjif — CHANGES_REQUESTED** [review]
The code looks good but after a little but when Genesis Plaza loads, we get this error in the console and the automatic doors from the ground floor don't work, they never open and you can get trapped in the main building:
Screen Shot 2022-03-29 at 15 05 30

**Comment by pbosio** [comment]
> The code looks good but after a little but when Genesis Plaza loads, we get this error in the console and the automatic doors from the ground floor don't work, they never open and you can get trapped in the main building: Screen Shot 2022-03-29 at 15 05 30

fixed. seems like I broke something with quests when I did some changes to add tests

**Review by pravusjif — APPROVED** [review]

**Pull request merged**